### PR TITLE
LB-628: Add support for incremental dumps

### PR DESCRIPTION
**Pull request by ishaanshah**
# Problem
The frequency at which stats are updated is very less.

# Solution
Adding support for incremental dumps will solve the issue.

### CODE REVIEW

File: listenbrainz_spark/hdfs/__init__.py [hunk 1]
```diff
@@ -62,6 +62,17 @@ def upload_archive(self, tmp_dump_dir, tar, dest_path, schema, callback=None, fo
         if callback is None:
             raise NotImplementedError('Callback to process JSON missing. Aborting...')
 
+        # Delete '/temp'if it exists
+        if utils.path_exists("/temp"):
```
Comment (paramsingh) [inline listenbrainz_spark/hdfs/__init__.py:66]: the `/temp` dir here should be a CONSTANT up at the top of the file imo, it's used in a bunch of places.

File: listenbrainz_spark/utils.py [hunk 2]
```diff
@@ -311,3 +311,23 @@ def rename(hdfs_src_path: str, hdfs_dst_path: str):
...
+
+def copy(hdfs_src_path: str, hdfs_dst_path: str, overwrite: bool = False):
+    """ Copy a file or folder in HDFS
+
+        Args:
+            hdfs_src_path – Source path.
+            hdfs_dst_path – Destination path. If the path already exists and is a directory, the source will be copied into it.
+            overwrite - Wether to overwrite the path if it already exists.
+    """
+    print(hdfs_src_path)
```
Comment (paramsingh) [inline listenbrainz_spark/utils.py:324]: this is extraneous.

File: listenbrainz_spark/utils.py [hunk 3]
```diff
@@ -311,3 +311,23 @@ def rename(hdfs_src_path: str, hdfs_dst_path: str):
...
+    walk = hdfs_walk(hdfs_src_path)
+
+    for (root, dirs, files) in walk:
+        for _file in files:
+            src_file_path = os.path.join(root, _file)
+            dst_file_path = os.path.join(hdfs_dst_path, src_file_path.replace(hdfs_src_path, '', 1).strip("/"))
```
Comment (paramsingh) [inline listenbrainz_spark/utils.py:330]: what's the replace here for? Maybe this should be using the os.path functions instead?

File: listenbrainz_spark/hdfs/upload.py [hunk 4]
```diff
@@ -46,43 +47,48 @@ def process_json_listens(self, filename, data_dir, tmp_hdfs_path, schema):
             month = filename.split('/')[-1][0:-5]
             dest_path = os.path.join(data_dir, year, '{}.parquet'.format(str(month)))
 
+        if utils.path_exists(dest_path):
```
Comment (paramsingh) [inline listenbrainz_spark/hdfs/upload.py:50]: This doesn't make sense in the first read, I figure we should have a parameter passed to the function which tells it to append instead.

File: listenbrainz_spark/hdfs/upload.py [hunk 5]
```diff
@@ -46,43 +47,48 @@ def process_json_listens(self, filename, data_dir, tmp_hdfs_path, schema):
...
+                self.upload_archive(tmp_dump_dir, tar, path.MBID_MSID_MAPPING, schema.msid_mbid_mapping_schema,
+                                    self.process_json, force=True)
 
     def upload_listens(self, archive, force=False):
         """ Decompress archive and upload listens to HDFS.
 
             Args:
                 archive: listens tar file to upload.
-                force: If True deletes dir at path where listens are to be uploaded.
+                force: If True deletes dir at path where listens are to be uploaded,
```
Comment (paramsingh) [inline listenbrainz_spark/hdfs/upload.py:75]: We should consider renaming force here to `incremental` or something like that

**Review by paramsingh — CHANGES_REQUESTED** [review]

File: listenbrainz_spark/utils.py (hunk 2 above)
Comment (ishaanshah) [reply listenbrainz_spark/utils.py:324]: Removed

File: listenbrainz_spark/hdfs/__init__.py (hunk 1 above)
Comment (ishaanshah) [reply listenbrainz_spark/hdfs/__init__.py:66]: Added it

File: listenbrainz_spark/hdfs/upload.py (hunk 4 above)
Comment (ishaanshah) [reply listenbrainz_spark/hdfs/upload.py:50]: Added

File: listenbrainz_spark/hdfs/upload.py (hunk 5 above)
Comment (ishaanshah) [reply listenbrainz_spark/hdfs/upload.py:75]: Renamed it to `overwrite`

File: listenbrainz_spark/utils.py (hunk 3 above)
Comment (ishaanshah) [reply listenbrainz_spark/utils.py]: Changed to `os.path.relpath`.

File: listenbrainz/spark/handlers.py [hunk 6]
```diff
@@ -197,7 +196,7 @@ def handle_dump_imported(data):
     dump_name = data['imported_dump']
     import_completion_time = data['time']
     send_mail(
-        subject='A full data dump has been imported into the Spark cluster',
+        subject='A data dump has been imported into the Spark cluster',
```
Comment (paramsingh) [inline listenbrainz/spark/handlers.py:199]: can we put in info on what type of dump this was in the subject as well? it's helpful when searching through the inbox.

**Review by paramsingh — APPROVED** [review]

**Pull request merged**